oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSourceSchemaReader.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.oracle;
+
+import com.google.common.collect.ImmutableSet;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.plugin.db.CommonSchemaReader;
+
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.Set;
+
+/**
+ * Oracle Source schema reader.
+ */
+public class OracleSourceSchemaReader extends CommonSchemaReader {
+  /**
+   * Oracle type constants, from Oracle JDBC Implementation.
+   */
+  public static final int INTERVAL_YM = -103;
+  public static final int INTERVAL_DS = -104;
+  public static final int TIMESTAMP_TZ = -101;
+  public static final int TIMESTAMP_LTZ = -102;
+  public static final int BINARY_FLOAT = 100;
+  public static final int BINARY_DOUBLE = 101;
+  public static final int BFILE = -13;
+  public static final int LONG = -1;
+  public static final int LONG_RAW = -4;
+
+  public static final Set<Integer> ORACLE_TYPES = ImmutableSet.of(
+    INTERVAL_DS,
+    INTERVAL_YM,
+    TIMESTAMP_TZ,
+    TIMESTAMP_LTZ,
+    BINARY_FLOAT,
+    BINARY_DOUBLE,
+    BFILE,
+    LONG,
+    LONG_RAW,
+    Types.NUMERIC,
+    Types.DECIMAL
+  );
+
+  @Override
+  public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLException {
+    int sqlType = metadata.getColumnType(index);
+
+    switch (sqlType) {
+      case TIMESTAMP_TZ:
+        return Schema.of(Schema.Type.STRING);
+      case TIMESTAMP_LTZ:
+        return Schema.of(Schema.LogicalType.TIMESTAMP_MICROS);
+      case BINARY_FLOAT:
+        return Schema.of(Schema.Type.FLOAT);
+      case BINARY_DOUBLE:
+        return Schema.of(Schema.Type.DOUBLE);
+      case BFILE:
+      case LONG_RAW:
+        return Schema.of(Schema.Type.BYTES);
+      case INTERVAL_DS:
+      case INTERVAL_YM:
+      case LONG:
+        return Schema.of(Schema.Type.STRING);
+      case Types.NUMERIC:
+      case Types.DECIMAL:
+        // This is the only way to differentiate FLOAT/REAL columns from other numeric columns, that based on NUMBER.
+        // Since in Oracle FLOAT is a subtype of the NUMBER data type, 'getColumnType' and 'getColumnTypeName' can not
+        // be used.
+        if (Double.class.getTypeName().equals(metadata.getColumnClassName(index))) {
```

Comment:
`FLOAT`/`REAL` columns are also reported as `Types.NUMERIC` by JDBC driver and this is the only way to differentiate these columns from other numeric columns. 

`metadata.getColumnClassName` returns `java.lang.Double` for these columns, whereas `java.math.BigDecimal` returned for other numeric columns. 

These are mapped to Double since they can store values greater than `Float.MAX_VALUE` and Oracle's JDBC driver returns `java.lang.Double` if we read these columns via `ResultSet#getObject`.

********************************************************************

> Which oracle type becomes a double?

`BINARY_DOUBLE`, `FLOAT`, `REAL`

Reply:
I see. It wasn't clear to me from the comment. Can you modify it to say the BINARY_DOUBLE, FLOAT, and REAL are returned as SQL Types.NUMERIC, but with value that is a java Double?

Reply:
Sorry, I misunderstood the question. FLOAT/REAL are returned as java.sql.Types.NUMERIC and BINARY_DOUBLE is returned as custom Oracle type:
```
public static final int BINARY_DOUBLE = 101;
```

Comment changed to:
```
...
// FLOAT and REAL are returned as java.sql.Types.NUMERIC but with value that is a java.lang.Double
if (Double.class.getTypeName().equals(metadata.getColumnClassName(index))) {
...
```